File: app/src/main/java/com/elitec/alejotaller/feature/auth/domain/util/ProfileClassification.kt
package com.elitec.alejotaller.feature.auth.domain.util

import com.elitec.alejotaller.feature.auth.domain.entity.User

/**
 * Mirrors web AUTH_POLICY / profile-classification.ts
 * If the profile is not clearly authenticated → visitor.
 */
enum class SessionMode {
    AUTHENTICATED,
    VISITOR
}

private val GUEST_PROVIDERS = setOf("guest", "anonymous", "visitor", "invitado")

fun isGuestProvider(provider: String?): Boolean {
    if (provider.isNullOrBlank()) return false
    return provider.trim().lowercase() in GUEST_PROVIDERS
}

fun isAnonymousProfile(email: String?, provider: String? = null): Boolean {
    if (isGuestProvider(provider)) return true
    return email.isNullOrBlank()
}

fun hasClearAuthenticatedProfile(
    id: String?,
    email: String?,
    provider: String? = null
): Boolean {
    if (id.isNullOrBlank()) return false
    if (isGuestProvider(provider)) return false
    if (isAnonymousProfile(email, provider)) return false
    return !email.isNullOrBlank()
}

fun hasClearAuthenticatedProfile(user: User?): Boolean {
    if (user == null) return false
    return hasClearAuthenticatedProfile(id = user.id, email = user.email, provider = null)
}

fun classifySessionMode(user: User?): SessionMode {
    return if (hasClearAuthenticatedProfile(user)) SessionMode.AUTHENTICATED else SessionMode.VISITOR
}

fun classifySessionMode(id: String?, email: String?, provider: String? = null): SessionMode {
    return if (hasClearAuthenticatedProfile(id, email, provider)) {
        SessionMode.AUTHENTICATED
    } else {
        SessionMode.VISITOR
    }
}

File: web/src/core/feature/auth/presentation/util/profile-classification.ts
import { isAnonymousAppwriteUser, isGuestProvider } from "./gest-session";

/**
 * Classifies the current Appwrite account into a clear session mode.
 * Policy: if the profile is not clearly authenticated, treat as visitor (guest).
 *
 * Clear authenticated profile requires a non-empty email (or explicit non-guest provider).
 * Anonymous / empty-email / guest-provider accounts are always visitors.
 */
export type SessionMode = "authenticated" | "visitor";

export type ClassifiableUser = {
    id?: string | null;
    $id?: string | null;
    email?: string | null;
    name?: string | null;
    provider?: string | null;
    role?: string | null;
} | null | undefined;

export function hasClearAuthenticatedProfile(user: ClassifiableUser): boolean {
    if (!user) return false;

    const id = (user.id ?? user.$id ?? "").toString().trim();
    if (!id) return false;

    if (isGuestProvider(user.provider)) return false;
    if (isAnonymousAppwriteUser(user)) return false;

    const email = typeof user.email === "string" ? user.email.trim() : "";
    // Clear client/admin profile must have an email identity
    if (!email) return false;

    return true;
}

export function classifySessionMode(user: ClassifiableUser): SessionMode {
    return hasClearAuthenticatedProfile(user) ? "authenticated" : "visitor";
}

export function resolveUserId(user: ClassifiableUser): string | null {
    if (!user) return null;
    const id = (user.id ?? user.$id ?? "").toString().trim();
    return id || null;
}
